### PR TITLE
Add draft of release guide document

### PR DESCRIPTION
**Pull request by pchlap**
@SimonBiggs & @Matthew-Jennings, 

0.39.0 is live! 🥳 

Here is the document I have put together as I worked through this. Please let me know if you have any suggested changes/clarifications. Not rush on this though.

One issue I did notice with 0.39.0, the final sanity check requires running the tests using the pymedphys tool: `pymedphys dev tests`. Likely due to some rejigging of the extras, the advice is to install pymedphys using `pip install pymedphys[user]`. However this didn't install what I needed to run the tests, I had to run `pip install pymedphys[all]`.

In addition the `test_lossless_jpeg` test is failing for me locally (running on MacOS). Could this be due to the pylibjpeg-libjpeg version?

Let me know if we should separate this discussion out to an issue.

### CODE REVIEW

**Comment by Matthew-Jennings** [comment]
Thank you so much, @pchlap ! I've given it a first pass and I think this looks great!

Just a question (perhaps for @SimonBiggs): I notice that each version of `pymedphys` has been uploaded to Zendodo - is this automatic, or do I need to do this manually for the purposes of the JOSS review?

File: lib/pymedphys/docs/contrib/info/release-guide.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+### Publish release on GitHub
+
+Create a new release on [GitHub](https://github.com/pymedphys/pymedphys/releases). Create a new tag in the format `vMAJOR.MINOR.PATCH` (e.g. `v0.39.0`). Enter the tag as the name for this release as well. Finally add in the change logs for this release (direct copy from CHANGELOG.md). Publish the release.
+
+### Final sanity tests
+
+Perform a final check to ensure the new version was released successfully. To do this, create a fresh Python virtual environment on your machine. Then install the new version of pymedphys and ensure all tests are passing as expected.
+
+```python
+pip install pymedphys[all]
```
Comment (SimonBiggs) [inline lib/pymedphys/docs/contrib/info/release-guide.md:109]:
```suggestion
pip install pymedphys[user,tests]
```

**Review by SimonBiggs — APPROVED** [review]
This is beautiful stuff @pchlap! Thank you!

**Comment by pchlap** [comment]
> Yup, by design. If you run `pip install pymedphys[user,tests]` that should be sufficient.

OK great, the only thing is the error message to the user says to simply install `pip install pymedphys[user]`, is this intended?

**Comment by SimonBiggs** [comment]
> > Yup, by design. If you run `pip install pymedphys[user,tests]` that should be sufficient.
> 
> OK great, the only thing is the error message to the user says to simply install `pip install pymedphys[user]`, is this intended?

Yeah... that should cover most use cases. It potentially could be worth some further logic to detect if the user needs to install other optional dependencies also... But... if a user is "using" the pymedphys library then that install message should cover all of those use cases.

...if they are testing, then `pymedphys[user]` is not enough. 

The fix for this could be, to have a dependency list created from the requirements files that is saved into the python package via propagate, and then read out the following `modname` variable:

https://github.com/pymedphys/pymedphys/blob/778e33769ead3ed6a6b366fd816f59aa8a2d8e76/lib/pymedphys/_vendor/apipkg/__init__.py#L248-L256

And then, if that `modname` variable is within the `user` dependency list, then ask the user to run `pip install pymedphys[user]`. If it is in the tests dependency list, then ask them to run `pip install pymedphys[user,tests]`...

File: lib/pymedphys/docs/contrib/info/release-guide.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+Then run poetry update as well as propagate:
+
+```bash
+poetry update
+poetry run pymedphys propagate
+```
+
+## Update CHANGELOG
+
+Amend the `CHANGELOG.md` file to describe the changes since the last release. Insert information for this release near the top of the file and populate the sections as appropriate (remove unused sections):
```
Comment (Matthew-Jennings) [inline lib/pymedphys/docs/contrib/info/release-guide.md:41]: Doesn't affect what you've written here, but I think it makes sense to enforce CHANGELOG edits at the time of contribution. We can add this is a future "how to make a contribution" page

File: lib/pymedphys/docs/contrib/info/release-guide.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+Commit all changes to your release branch and push to GitHub. Then create a Pull Request of this branch into main. This will run the CI tests. Before finalising the release, ensure all tests are passing.
+
+### Troubleshooting issues
+
+When preparing a release you will typically see certain steps of the CI failing. This is usually due to dependencies having been upgraded in the time since the last release and certain code within pymedphys is no longer compatible with the new versions.
+
+To confirm this is indeed the problem, you can set an upper pin for these libraries (to constrain them to a previous version). By doing this you should hopefully see the CI pass.
+
+In most instances, releasing pymedphys with an upper pin isn't appropriate, as this would be too restrictive for pymedphys users. Unfortunately there isn't a general solution to resolving these issues. You will need to remove the upper pins one-by-one and adjust the code in pymedphys to support the new version of the dependencies.
```
Comment (Matthew-Jennings) [inline lib/pymedphys/docs/contrib/info/release-guide.md:78]:
```suggestion
In most instances, releasing pymedphys with an upper pin isn't appropriate. `pymedphys` is primarily a _library_ and upper pins introduce future obsolescnece risk. Unfortunately there isn't a general solution to resolving these issues. You will need to remove the upper pins one-by-one and adjust the code in pymedphys to support the new version of the dependencies.
```

**Review by Matthew-Jennings — APPROVED** [review]

File: lib/pymedphys/docs/contrib/info/release-guide.md (hunk 2 above)
Comment (SimonBiggs) [reply lib/pymedphys/docs/contrib/info/release-guide.md:41]: GitHub has a PR checklist option. @Matthew-Jennings, would you be open to setting that up and including a checkbox in the template for "added relevant changes to the changelog"?

File: lib/pymedphys/docs/contrib/info/release-guide.md (hunk 3 above)
Comment (SimonBiggs) [reply lib/pymedphys/docs/contrib/info/release-guide.md:78]: Less so obsolescence... more so... "dependency hell" for those trying to install the library in combination with other libraries that have their own restrictions.

Comment (Matthew-Jennings) [reply lib/pymedphys/docs/contrib/info/release-guide.md:78]: Sure, "primarily" misses the point that this is a library release.

I do think upper pins technically introduce obsolescence, which at least worsens in the "future", but I can see that perhaps it's not the primary point. Feel free to suggest alternate wording. I just thought it best to give a reason.

I had in mind the point that, at some time in the future, a pymedphys dependency will likely introduce lower-pins that conflict with another (cross-)dependency's upper-pin that we've introduced. So, "obsolescence" and "future".

File: lib/pymedphys/docs/contrib/info/release-guide.md (hunk 2 above)
Comment (Matthew-Jennings) [reply lib/pymedphys/docs/contrib/info/release-guide.md:41]: Yep, after around 10th October, I can get stuck into that. I'm going to be slammed between now and then.

**Comment by SimonBiggs** [comment]
Absolutely beautiful stuff @pchlap! A key goal for PyMedPhys has been to be able to have it be able to be built and run by us all. A release where I didn't touch it :slightly_smiling_face:! That's a first. Thank you :slightly_smiling_face: :tada:

**Pull request merged**